src/libraries/System.Threading/tests/ExecutionContextTests.cs

```diff
@@ -30,6 +30,27 @@ public static void CreateCopyTest()
             });
         }
 
+        [Fact]
+        public static void RestoreTest()
+        {
+            ExecutionContext defaultEC = ExecutionContext.Capture();
+            var asyncLocal = new AsyncLocal<int>();
+            Assert.Equal(0, asyncLocal.Value);
+
+            asyncLocal.Value = 1;
+            ExecutionContext oneEC = ExecutionContext.Capture();
+            Assert.Equal(1, asyncLocal.Value);
+
+            ExecutionContext.Restore(defaultEC);
+            Assert.Equal(0, asyncLocal.Value);
+
+            ExecutionContext.Restore(oneEC);
+            Assert.Equal(1, asyncLocal.Value);
+
+            ExecutionContext.Restore(defaultEC);
+            Assert.Equal(0, asyncLocal.Value);
+        }
```

Comment:
Can you add tests for null?

Reply:
Ah, hmm...

So was going to change it to throw on `null`; but that means can't restore a flowsupressed context (as that's null).

However allowing `null` also means can't restore a flow suppressed context with any asynclocals set (as it is captured as `null`)

Fine for the use case I want it for; but not sure what to do for the general api?

e.g. this would be invalid in `ManualResetValueTaskSourceCore.cs` as it would blank a flow suppressed context back to having no asynclocals? https://github.com/dotnet/runtime/blob/fe81abd171228e738bdea44aedb86acc9b32f239/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs#L259-L263

Reply:
So MRVTSC is now broken in the face of flow suppression? If yes, wet need to revert that change until this is solved.

Reply:
It's easy enough to fix internally; but then don't allow `null` contexts externally? (as per EC.Run)

Reply:
Why wouldn't external consumption have the same issue?

Reply:
Fixed

Reply:
> Why wouldn't external consumption have the same issue?

Now externally it won't let you restore a flow suppressed context (like you can't `.Run` on one); so the behaviour is consistent.

If you wanted to `.Restore` one, you have to do a little dance of switching off suppression, capturing, then switching it back on after restore e.g.

```csharp
ExecutionContext.RestoreFlow()
var ec = ExecutionContext.Capture()
ExecutionContext.SuppressFlow();

...

ExecutionContext.Restore(ec)
ExecutionContext.SuppressFlow();
```